README.md
# Iptables Log Parser

### Overview
Iptables is a utility program that allows a system administrator to configure the IP packet filter rules of the Linux kernel firewall, implemented as different Netfilter modules. The filters are organized in different tables, which contain chains of rules for how to treat network traffic packets. This repository serves to parse an Iptables log. 

### Technology
Haskell

CHANGELOG.md
# Revision history for iptables-log-parser

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
{-# language PatternSynonyms #-}

import Data.Bytes (Bytes)
import Linux.Iptables (Attribute(..),decode)
import Net.Protocol (pattern Tcp)

import qualified Data.Bytes as Bytes
import qualified Net.IPv4 as IPv4
import qualified Net.Mac as Mac

main :: IO ()
main = do
  putStrLn "Starting"
  testA
  putStrLn "Finished"

testA :: IO ()
testA = case decode sampleA of
  Nothing -> fail "Could not decode A"
  Just attrs
    | notElem (DestinationPort 35720) attrs -> fail "bad destination port"
    | notElem (SourcePort 443) attrs -> fail "bad source port"
    | notElem (Destination (IPv4.fromOctets 192 168 1 10)) attrs -> fail "bad dest ip"
    | notElem (Source (IPv4.fromOctets 108 177 122 189)) attrs -> fail "bad source ip"
    | notElem (Protocol Tcp) attrs -> fail "bad protocol"
    | notElem (InInterface (Bytes.fromLatinString "wlp2s0")) attrs -> fail "bad in interface"
    | notElem (Mac (Mac.mac 0xf48c503e8eba) (Mac.mac 0x08028ee7f6a4)) attrs -> fail "bad MAC addresses"
    | otherwise -> pure ()

sampleA :: Bytes
sampleA = Bytes.fromLatinString $ concat
  [ "Apr 30 08:04:42 thadtop kernel: [50514.038553] IN=wlp2s0 OUT= "
  , "MAC=f4:8c:50:3e:8e:ba:08:02:8e:e7:f6:a4:08:00 SRC=108.177.122.189 "
  , "DST=192.168.1.10 LEN=105 TOS=0x00 PREC=0x20 TTL=119 ID=45030 PROTO=TCP "
  , "SPT=443 DPT=35720 WINDOW=830 RES=0x00 ACK PSH URGP=0"
  ]



src/Net/Protocol.hs
{-# language DerivingStrategies #-}
{-# language GeneralizedNewtypeDeriving #-}
{-# language MagicHash #-}
{-# language PatternSynonyms #-}
{-# language TypeApplications #-}

module Net.Protocol
  ( -- * Types
    Protocol(..)
    -- * Decode
  , decode
    -- * Decode
  , hashMap
    -- * Patterns
  , pattern Tcp
  , pattern Igp
  , pattern Udp
  , pattern Sctp
  , pattern Icmp
  , pattern Igmp
  ) where

import Data.Bytes (Bytes)
import Data.Bytes.HashMap.Word (Map)
import Data.Char (toLower,toUpper)
import Data.Word (Word8)
import GHC.Base (unpackCString#)
import GHC.Exts (Addr#)

import qualified Data.Bytes as Bytes
import qualified Data.Bytes.HashMap.Word as Map

newtype Protocol = Protocol Word8
  deriving newtype (Eq)

pattern Tcp :: Protocol
pattern Tcp = Protocol 6

pattern Igp :: Protocol
pattern Igp = Protocol 9

pattern Udp :: Protocol
pattern Udp = Protocol 17

pattern Sctp :: Protocol
pattern Sctp = Protocol 132

pattern Icmp :: Protocol
pattern Icmp = Protocol 1

pattern Igmp :: Protocol
pattern Igmp = Protocol 2

decode :: Bytes -> Maybe Protocol
decode b = case Map.lookup b hashMap of
  Nothing -> Nothing
  Just w -> Just (Protocol (fromIntegral w))

-- | Resolves protocol names to IANA numbers. 
hashMap :: Map
hashMap = Map.fromTrustedList $ do
  RawPair key val <- rawProtocolPairs
  let val' = fromIntegral @Word8 @Word val
      str = unpackCString# key
      upper = Bytes.fromLatinString (map toUpper str)
      lower = Bytes.fromLatinString (map toLower str)
  [(upper, val'), (lower, val'), (Bytes.fromLatinString str, val')]

data RawPair = RawPair Addr# !Word8

rawProtocolPairs :: [RawPair]
rawProtocolPairs =
  [ RawPair "HOPOPT"# 0
  , RawPair "ICMP"# 1
  , RawPair "IGMP"# 2
  , RawPair "GGP"# 3
  , RawPair "IPv4"# 4
  , RawPair "ST"# 5
  , RawPair "TCP"# 6
  , RawPair "CBT"# 7
  , RawPair "EGP"# 8
  , RawPair "IGP"# 9
  , RawPair "BBN-RCC-MON"# 10
  , RawPair "NVP-II"# 11
  , RawPair "PUP"# 12
  , RawPair "EMCON"# 14
  , RawPair "XNET"# 15
  , RawPair "CHAOS"# 16
  , RawPair "UDP"# 17
  , RawPair "MUX"# 18
  , RawPair "DCN-MEAS"# 19
  , RawPair "HMP"# 20
  , RawPair "PRM"# 21
  , RawPair "XNS-IDP"# 22
  , RawPair "TRUNK-1"# 23
  , RawPair "TRUNK-2"# 24
  , RawPair "LEAF-1"# 25
  , RawPair "LEAF-2"# 26
  , RawPair "RDP"# 27
  , RawPair "IRTP"# 28
  , RawPair "ISO-TP4"# 29
  , RawPair "NETBLT"# 30
  , RawPair "MFE-NSP"# 31
  , RawPair "MERIT-INP"# 32
  , RawPair "DCCP"# 33
  , RawPair "3PC"# 34
  , RawPair "IDPR"# 35
  , RawPair "XTP"# 36
  , RawPair "DDP"# 37
  , RawPair "IDPR-CMTP"# 38
  , RawPair "TP++"# 39
  , RawPair "IL"# 40
  , RawPair "IPv6"# 41
  , RawPair "SDRP"# 42
  , RawPair "IPv6-Route"# 43
  , RawPair "IPv6-Frag"# 44
  , RawPair "IDRP"# 45
  , RawPair "RSVP"# 46
  , RawPair "GRE"# 47
  , RawPair "DSR"# 48
  , RawPair "BNA"# 49
  , RawPair "ESP"# 50
  , RawPair "AH"# 51
  , RawPair "I-NLSP"# 52
  , RawPair "SWIPE"# 53
  , RawPair "NARP"# 54
  , RawPair "MOBILE"# 55
  , RawPair "TLSP"# 56
  , RawPair "SKIP"# 57
  , RawPair "IPv6-ICMP"# 58
  , RawPair "IPv6-NoNxt"# 59
  , RawPair "IPv6-Opts"# 60
  , RawPair "CFTP"# 62
  , RawPair "SAT-EXPAK"# 64
  , RawPair "KRYPTOLAN"# 65
  , RawPair "RVD"# 66
  , RawPair "IPPC"# 67
  , RawPair "SAT-MON"# 69
  , RawPair "VISA"# 70
  , RawPair "IPCV"# 71
  , RawPair "CPNX"# 72
  , RawPair "CPHB"# 73
  , RawPair "WSN"# 74
  , RawPair "PVP"# 75
  , RawPair "BR-SAT-MON"# 76
  , RawPair "SUN-ND"# 77
  , RawPair "WB-MON"# 78
  , RawPair "WB-EXPAK"# 79
  , RawPair "ISO-IP"# 80
  , RawPair "VMTP"# 81
  , RawPair "SECURE-VMTP"# 82
  , RawPair "VINES"# 83
  , RawPair "TTP"# 84
  , RawPair "IPTM"# 84
  , RawPair "NSFNET-IGP"# 85
  , RawPair "DGP"# 86
  , RawPair "TCF"# 87
  , RawPair "EIGRP"# 88
  , RawPair "OSPFIGP"# 89
  , RawPair "Sprite-RPC"# 90
  , RawPair "LARP"# 91
  , RawPair "MTP"# 92
  , RawPair "AX.25"# 93
  , RawPair "IPIP"# 94
  , RawPair "MICP"# 95
  , RawPair "SCC-SP"# 96
  , RawPair "ETHERIP"# 97
  , RawPair "ENCAP"# 98
  , RawPair "GMTP"# 100
  , RawPair "IFMP"# 101
  , RawPair "PNNI"# 102
  , RawPair "PIM"# 103
  , RawPair "ARIS"# 104
  , RawPair "SCPS"# 105
  , RawPair "QNX"# 106
  , RawPair "A/N"# 107
  , RawPair "IPComp"# 108
  , RawPair "SNP"# 109
  , RawPair "Compaq-Peer"# 110
  , RawPair "IPX-in-IP"# 111
  , RawPair "VRRP"# 112
  , RawPair "PGM"# 113
  , RawPair "L2TP"# 115
  , RawPair "DDX"# 116
  , RawPair "IATP"# 117
  , RawPair "STP"# 118
  , RawPair "SRP"# 119
  , RawPair "UTI"# 120
  , RawPair "SMP"# 121
  , RawPair "SM"# 122
  , RawPair "PTP"# 123
  , RawPair "ISIS over IPv4"# 124
  , RawPair "FIRE"# 125
  , RawPair "CRTP"# 126
  , RawPair "CRUDP"# 127
  , RawPair "SSCOPMCE"# 128
  , RawPair "IPLT"# 129
  , RawPair "SPS"# 130
  , RawPair "PIPE"# 131
  , RawPair "SCTP"# 132
  , RawPair "FC"# 133
  , RawPair "RSVP-E2E-IGNORE"# 134
  , RawPair "Mobility Header"# 135
  , RawPair "UDPLite"# 136
  , RawPair "MPLS-in-IP"# 137
  , RawPair "manet"# 138
  , RawPair "HIP"# 139
  , RawPair "Shim6"# 140
  , RawPair "WESP"# 141
  , RawPair "ROHC"# 142
  , RawPair "Ethernet"# 143
  ]


src/Linux/Iptables.hs
{-# language BangPatterns #-}
{-# language DerivingStrategies #-}
{-# language LambdaCase #-}

module Linux.Iptables
  ( Attribute(..)
  , decode
  ) where

-- Note: I played around with stream fusion in here to see if I could get GHC
-- to generate fast parsing code without having so much boilerplate. I eventually
-- got it to do what I wanted, but it was pretty difficult, and I had to manually
-- fuse two of the passes and get rid of some error checking. I would not recommend
-- using stream fusion for this kind of thing. Also, there is a phase-restricted
-- inline pragma on tokenizeAndParse, and if you remove it, it causes the compiler
-- to hang. I don't know why.

import Data.Bytes.Types (Bytes(Bytes))
import Data.Bytes.Parser (Parser)
import Data.Word (Word16)
import Net.Types (IPv4)
import Data.Bytes.HashMap (Map)
import Net.Protocol (Protocol)
import Data.Word (Word8)

import qualified Data.Bytes as Bytes
import qualified Data.Bytes.HashMap as Map
import qualified Data.Vector.Fusion.Stream.Monadic as Stream
import qualified Data.Bytes.Parser as Parser
import qualified Data.Bytes.Parser.Latin as Latin
import qualified Net.IPv4 as IPv4
import qualified Net.Mac as Mac
import qualified Net.Protocol as Protocol
import qualified Net.Types as Net

data Attribute
  = Source !IPv4
  | Destination !IPv4
  | SourcePort !Word16
  | DestinationPort !Word16
  | Length !Word16
  | TimeToLive !Word8
  | InInterface {-# UNPACK #-} !Bytes
  | OutInterface {-# UNPACK #-} !Bytes
  | Mac !Net.Mac !Net.Mac
    -- ^ For some reason, iptables logs put the source MAC address and the
    -- destination MAC address together. We tease them apart. The first one
    -- is the source, and the second one is the destination.
  | Protocol !Protocol -- ^ @PROTO@
  | Acknowledgement -- ^ @ACK@
  | Synchronize -- ^ @SYN@
  deriving stock (Eq)

data Field
  = FieldSource
  | FieldDestination
  | FieldSourcePort
  | FieldDestinationPort
  | FieldProtocol
  | FieldLength
  | FieldInInterface
  | FieldOutInterface
  | FieldMac

data Tag
  = TagAcknowledgement
  | TagSynchronize

decode :: Bytes -> Maybe [Attribute]
decode b@(Bytes arr _ _)
    -- Drop everything until the first equals sign.
  | !b'@(Bytes _ off' len') <- Bytes.dropWhile (\c -> c /= 0x3D) b
  , Bytes.length b >= Bytes.length b' + 2
  , adjusted <- Bytes arr (off' - 2) (len' - 2)
    -- Drop trailing spaces, carriage returns, and newlines. These
    -- often end up at the end of logs.
  , trimmed <- Bytes.dropWhileEnd (\c -> c == 0x0A || c == 0x0D || c == 0x20) adjusted =
    -- These wildcard bangs are here so stop GHC from repeatedly
    -- forcing the hashmaps while parsing the fields.
    let !_ = tags
        !_ = fields
        !_ = Protocol.hashMap
        -- If you specialize the monad to Identity instead of Just, the
        -- generated code becomes way worse. Not sure why.
     in Stream.foldl' (flip (:)) []
      . Stream.mapMaybe tokenizeAndParse
      . Bytes.splitStream 0x20
      $ trimmed

fields :: Map Field
fields = Map.fromTrustedList
  [ (Bytes.fromLatinString "SRC", FieldSource)
  , (Bytes.fromLatinString "DST", FieldDestination)
  , (Bytes.fromLatinString "SPT", FieldSourcePort)
  , (Bytes.fromLatinString "DPT", FieldDestinationPort)
  , (Bytes.fromLatinString "PROTO", FieldProtocol)
  , (Bytes.fromLatinString "LEN", FieldLength)
  , (Bytes.fromLatinString "IN", FieldInInterface)
  , (Bytes.fromLatinString "OUT", FieldOutInterface)
  , (Bytes.fromLatinString "MAC", FieldMac)
  ]

tags :: Map Tag
tags = Map.fromTrustedList
  [ (Bytes.fromLatinString "ACK", TagAcknowledgement)
  , (Bytes.fromLatinString "SYN", TagSynchronize)
  ]

tagToAttribute :: Tag -> Attribute
tagToAttribute = \case
  TagAcknowledgement -> Acknowledgement
  TagSynchronize -> Synchronize

fieldToAttribute :: Field -> Bytes -> Maybe Attribute
fieldToAttribute x !b = case x of
  FieldSource -> Source <$> IPv4.decodeUtf8Bytes b
  FieldDestination -> Destination <$> IPv4.decodeUtf8Bytes b
  FieldSourcePort -> decodeWord16 SourcePort b
  FieldDestinationPort -> decodeWord16 DestinationPort b
  FieldMac -> decodeMacPair b
  FieldProtocol -> Protocol <$> Protocol.decode b
  FieldLength -> decodeWord16 Length b
  FieldInInterface -> Just $ InInterface b
  FieldOutInterface -> Just $ OutInterface b

-- If the key is not recognized, we just skip it. Also, if we see
-- a key=val style field and the val is empty, just skip it. This
-- happens with interface names sometimes, and it is better to
-- just not report the field in this situation.
tokenizeAndParse :: Bytes -> Maybe Attribute
{-# inline [0] tokenizeAndParse #-}
tokenizeAndParse atom = case Bytes.split1 0x3D atom of
  Nothing -> case Map.lookup atom tags of
    Nothing -> Nothing
    Just tag -> Just (tagToAttribute tag)
  Just (key, b@(Bytes _ _ len)) -> case len of
    0 -> Nothing
    _ -> case Map.lookup key fields of
      Nothing -> Nothing
      Just field -> fieldToAttribute field b

parserPrefix :: Parser () s ()
parserPrefix = do
  skipInitialDate
  Latin.skipTrailedBy () '['
  Latin.skipDigits1 ()
  Latin.char () '.'
  Latin.skipDigits1 ()
  Latin.char () ']'

-- The initial datetime is formatted like this: Dec 6 10:04:50.
-- It is always missing the year. This consumes a trailing space.
skipInitialDate :: Parser () s ()
skipInitialDate = do
  match isUpper
  match isLower
  match isLower
  Latin.char () ' '
  Latin.skipDigits1 ()
  Latin.char () ' '
  match isDigit
  match isDigit
  Latin.char () ':'
  match isDigit
  match isDigit
  Latin.char () ':'
  match isDigit
  match isDigit
  Latin.char () ' '

isUpper :: Char -> Bool
isUpper c = c >= 'A' && c <= 'Z'

isLower :: Char -> Bool
isLower c = c >= 'a' && c <= 'z'

isDigit :: Char -> Bool
isDigit c = c >= '0' && c <= '9'

match :: (Char -> Bool) -> Parser () s ()
{-# inline match #-}
match p = do
  c <- Latin.any ()
  case p c of
    True -> pure ()
    False -> Parser.fail ()

decodeWord16 :: (Word16 -> x) -> Bytes -> Maybe x
decodeWord16 f b = Parser.parseBytesMaybe
  ( do w <- Latin.decWord16 ()
       pure (f w)
  ) b

decodeMacPair :: Bytes -> Maybe Attribute
decodeMacPair !b = Parser.parseBytesMaybe
  ( do src <- Mac.parserUtf8Bytes ()
       Latin.char () ':'
       dst <- Mac.parserUtf8Bytes ()
       Latin.char () ':'
       _ <- Latin.hexNibble ()
       _ <- Latin.hexNibble ()
       Latin.char () ':'
       _ <- Latin.hexNibble ()
       _ <- Latin.hexNibble ()
       pure (Mac src dst)
  ) b


bench/Main.hs
import Data.Bytes (Bytes)
import Gauge (bench,whnf,defaultMain)
import Linux.Iptables (decode)

import qualified Data.Bytes as Bytes

-- On my Intel(R) Xeon(R) CPU E3-1505M v5 @ 2.80GHz, this clocks in at
-- around 1.0 microseconds. To be honest, that is not great, but that
-- still puts us at around 1M logs per second. To do better, GHC probably
-- needs to get bytearray literals built in so that we can have the
-- compiler do the hash table lookup rather than doing them in library space.

main :: IO ()
main = defaultMain
  [ bench "A" (whnf decode sampleA)
  ]

sampleA :: Bytes
sampleA = Bytes.fromLatinString $ concat
  [ "Apr 30 08:04:42 thadtop kernel: [50514.038553] IN=wlp2s0 OUT= "
  , "MAC=f4:8c:50:3e:8e:ba:08:02:8e:e7:f6:a4:08:00 SRC=108.177.122.189 "
  , "DST=192.168.1.10 LEN=105 TOS=0x00 PREC=0x20 TTL=119 ID=45030 PROTO=TCP "
  , "SPT=443 DPT=35720 WINDOW=830 RES=0x00 ACK PSH URGP=0"
  ]

